mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+
+      let nameserver { nameserver ; _ } = nameserver
+      let rng { rng ; _ } = rng
+
+      let bind = Lwt.bind
+      let lift = Lwt.return
+
+      let cancel_all t =
+        IM.iter (fun _id cond ->
+            Lwt_condition.broadcast cond (Error (`Msg "disconnected")))
+          t.requests
+
+      let rec read_loop ?(linger = Cstruct.empty) t flow =
+        S.TCPV4.read flow >>= function
+        | Error e ->
+          t.flow <- None;
+          Log.err (fun m -> m "error %a reading from resolver" S.TCPV4.pp_error e);
+          cancel_all t;
+          Lwt.return_unit
+        | Ok `Eof ->
+          t.flow <- None;
+          Log.info (fun m -> m "end of file reading from resolver");
+          cancel_all t;
+          Lwt.return_unit
+        | Ok (`Data cs) ->
+          let rec handle_data data =
+            let cs_len = Cstruct.len data in
+            if cs_len > 2 then
+              let len = Cstruct.BE.get_uint16 data 0 in
+              if cs_len - 2 >= len then
+                let packet, rest = Cstruct.split data (len + 2) in
+                let id = Cstruct.BE.get_uint16 packet 2 in
+                (match IM.find_opt id t.requests with
+                 | None -> Log.warn (fun m -> m "received unsolicited data, ignoring")
+                 | Some cond ->
+                   inc `Recursive_answers;
+                   Lwt_condition.broadcast cond (Ok packet));
+                handle_data rest
+              else
+                read_loop ~linger:data t flow
+            else
+              read_loop ~linger:data t flow
+          in
+          handle_data (if Cstruct.len linger = 0 then cs else Cstruct.append linger cs)
+
+      let connect ?nameserver:ns t =
+        match t.flow with
+        | Some _ -> Lwt.return (Ok ({ t ; id = 0 }))
+        | None ->
+          let _proto, addr = match ns with None -> nameserver t | Some x -> x in
+          S.TCPV4.create_connection (S.tcpv4 t.stack) addr >|= function
+          | Error e ->
+            Log.err (fun m -> m "error connecting to nameserver %a"
+                        S.TCPV4.pp_error e) ;
+            Error (`Msg "connect failure")
+          | Ok flow ->
+            inc `Recursive_connections;
+            Lwt.async (fun () -> read_loop t flow);
+            t.flow <- Some flow;
+            Ok ({ t ; id = 0 })
+
+      let close _f =
+        (* ignoring this here *)
+        Lwt.return_unit
+
+      let recv { t ; id } =
+        let cond = Lwt_condition.create () in
+        t.requests <- IM.add id cond t.requests;
+        Lwt_condition.wait cond >|= fun data ->
+        t.requests <- IM.remove id t.requests;
+        match data with
+        | Ok cs -> Ok cs
+        | Error `Msg m -> Error (`Msg m)
+
+      let send f s =
+        match f.t.flow with
+        | None -> Lwt.return (Error (`Msg "no connection to resolver"))
+        | Some flow ->
+          let id = Cstruct.BE.get_uint16 s 2 in
```

Comment:
Is this always safe, ie do we need a guard on the length?

Reply:
yes, everything ocaml-dns creates is at least 12 bytes long